### DISCUSSION TITLE
Feature request: instant fixup without disturbing mod times.

### DISCUSSION THREAD

**Discussion by dabrahams** [magit/magit#5143] [Ideas]
Applying an instant fixup doesn't change the contents of the working tree, but in the projects I work on, it can be expensive for development because file modification times change.  It would be awesome if that operation could be done without touching the working tree files.

**Comment by kyleam** [comment]
In my view this feature would be outside the scope of Magit.  I could imagine a general non-Emacs tool that wires itself up to Git's hooks that and takes care of something like this.  It sounds like the case you're interested in (as well as other cases) would be covered by a tool that records the times the first time it sees a top-level tree and then restores those modification times if it sees that tree ID again after.

**Reply by dabrahams** [reply]
That tool would certainly be out of scope and huge overkill for a situation that only arises in practice when I use instant fixup. In fact the functionality as you described it would screw up build systems because reverting to an earlier state wouldn't cause files to be dirty.

Instant fixup is a magit-specific function and I'm just asking for it to be made better. It could do its work in a temporary worktree and reset HEAD.

**Reply by tarsius** [reply]
I agree with @kyleam.

**Reply by dabrahams** [reply]
@tarsius the tool @kyleam suggests doesn't work for reasons stated above.  So I surmise you're agreeing that an instant-fixup feature is in scope for Magit but doing it without forcing a recompile is out-of-scope?

I could understand “this is too much work; I can't do it,” but “out of scope” generally means “it's for use-cases we don't intend to address; I wouldn't even accept a patch.”  Is that what you mean?  I'm struggling to understand what use-case needs a fixup  that _does_ change mod times.  I would have thought magit is trying to address (among others) users that use compiled languages.

**Reply by tarsius** [reply]
Yes, I am not going to implement this and won't accept a patch (which going forward I would have to maintain).

I cannot implement and/or maintain everything that would be useful to someone.

**Reply by tarsius** [reply]
What is "out of scope" is not "doing something that only benefits static languages" but "doing something *inside* of Magit, which should be done *outside* of Magit [because it hasn't (yet?) been implemented elsewhere]".

One reason in this particular case is that it would be fairly simple to implement it in Git (or so I believe), while it would be difficult to implement in Magit and have unfortunate side-effects, such as increased runtime costs (which could be addressed by making it optional (which would have the unfortunate side-effects of increasing maintenance cost and further increasing complexity)).

So yes, "this is too much work; I can't do it," plays into it. But it is also very literally "out of scope" for Magit, because IMO it *should not be done in Magit*, but elsewhere.

**Reply by dabrahams** [reply]
My impression was that instant fixup itself is out of scope for GIt.  It's fine to implement it as a feature of a tool outside of git or even a `git-XXX` tool but then one would lose all the conveniences one gets from magit such as interactive target commit selection.  No need to respond to this; I just thought I should explain why I thought magit was the right home for this feature,

**Reply by tarsius** [reply]
The problem is that Magit implements "edit an older commit" basically doing just `git commit ...; git rebase ...`. Git then does it's thing, touching files in the process.

Magit isn't involved in that at all. It doesn't have to concern itself at all with determining what files are changed in the range of commits being replaced. Git does that and hopefully there are no conflicts. Only if there are conflicts (or when everything completed with no conflicts), does Magit get involved again. It looks in `.git/rebase-something` to learn if everything went well. It also does `git diff ...` as it always does when refreshing the status buffer, just in this case there may conflict markers.

But Magit is not involved in the fundling of files, and it cannot prevent Git from doing it and has no insight into which files are affected (the ones that end up with conflicts, sure, but other files likely are affected too).

Implementing this in Magit would mean recording the modification times of *all* files before the rebase, and then after rebase, check for changes and restore the old times. This is costly, certainly more costly than `git rebase` (optionally) simply not changing the modification time when touching a file.

Something that complicates this (whether implemented in Git, Magit or some third-party tool) is that the modification time actually has to be updated for files that do change. So it would not be enough to just record just the times, the tool also has to compare the initial state and end state, and then bump the modification times of anything that changed. Not hard, but if done in Magit it would no longer do basically just `git commit ...; git rebase ...`. Now it would also have to keep track of some state on its own, process it later and make changes to the file-system based on that on its own. This makes things more complicated by quite a bit. It's no longer just "git do your thing" and then waiting for the process to return.

**Reply by tarsius** [reply]
> My impression was that instant fixup itself is out of scope for GIt.

Magit's implementation involves `git rebase ...` and that is where modification times get modified. When `git rebase ...` is invoked directly by the user (whether on the command line or though Magit or some other third-party interface), the modification times or files whose contents do not change between the initial state and the post-rebase state are still changed, with the same consequences for the build system.

If Git implemented modtime-modification-avoidance for `git rebase`, Magit's "fixup" functionality would automatically benefit, and so would Magit's other uses of `git rebase`, uses on the command line and uses via other third-party interfaces. That's why I am saying this should be implemented in Git and why it is out of scope for Magit's fixup functionality.

**Reply by tarsius** [reply]
(

Also note that Git very recently gained some (still experimental and incomplete) functionality, which *might* solve this issue, for the very narrow use-case (rewording an older commit), that it actually already supports.

See https://github.com/magit/magit/discussions/5563 and go from there.

Magit does not support that yet. Eventually it probably will. Exactly because (i.e., assuming) it does not have the modtime issue. I haven't investigated yet and I will delay doing so until support in Git has matured (i.e., until it supports more than one or two use-cases, and actually begins to approach the feature set Magit has had for years). Even if this ends up doing everything Magit's instant fixup et al. does, I would not be able to through out the current `git rebase`-based implementation -- many users are going to stick to older Git releases for many years to come.

I.e.:

> My impression was that instant fixup itself is out of scope for GIt.

It seems that is no longer out of scope for Git, but they are just getting started.

)

**Reply by dabrahams** [reply]
> Magit isn't involved in that at all. It doesn't have to concern itself at all with determining what files are changed in the range of commits being replaced.

I realize that, which is why I suggested the worktree solution, which wouldn't require Magit to know that stuff (or look at mod times) either.  Again, I'm not insisting that you should implement any of these things, and as my programming life dwindles it won't make a huge difference to me either way; I just don't want you to think I was making a request without thinking of the level of complexity required from you. The worktree approach has its own complexities of course, though minor compared to trying to keep track of which files are being changed.

**Reply by dabrahams** [reply]
Wait, I don't see my worktree suggestion here.  I don't know what happened; could've sworn I made it.  The idea is more or less to create a temporary worktree, make the fixup there, then reset the HEAD of the regular working tree.  Plenty of little details involved that I didn't spell out here though.
